File: scripts/python/src/tools/utils/sysinfo/bench/runner.py
import os
import shutil
import subprocess
import sys
import time
from dataclasses import dataclass
from datetime import UTC, datetime

from tools.utils.sysinfo.bench import capture as environment
from tools.utils.sysinfo.bench.conditions import capture_conditions, gate_reasons, grade_for
from tools.utils.sysinfo.bench.limits import (
    COOLDOWN_INTERVAL,
    COOLDOWN_SECONDS,
    MAX_RUNS,
    MIN_RUNS,
    RSD_THRESHOLD,
    WRITE_BUDGET,
)
from tools.utils.sysinfo.bench.record import Metric, Run, epoch_of, relative_deviation
from tools.utils.sysinfo.bench.suites import (
    WRITTEN,
    MeasurementError,
    cpu,
    disk,
    gpu,
    memory,
    native,
    thermal,
    workload,
)
from tools.utils.sysinfo.collect import collect_snapshot

# thermal runs last because it saturates every core for up to two minutes.
# Ahead of workload it meant nvim startup and git status were timed on a hot,
# potentially clock-limited CPU -- the metrics most sensitive to thermal state.
SUITES = (cpu, native, memory, disk, gpu, workload, thermal)

FAMILIES = ("cpu", "mem", "cache", "disk", "gpu", "thermal", "workload")


class GateError(Exception):
    def __init__(self, reasons):
        super().__init__("; ".join(reasons))
        self.reasons = tuple(reasons)


@dataclass(frozen=True)
class Setting:
    tier: str
    workdir: str
    families: tuple[str, ...] = ()


def default_workdir():
    base = os.environ.get("XDG_CACHE_HOME") or os.path.expanduser("~/.cache")
    path = os.path.join(base, "dotfile", "bench", "work")
    os.makedirs(path, exist_ok=True)
    return path


def under(target, mount):
    return target == mount or target.startswith(mount.rstrip("/") + "/")


def darwin_filesystem(path):
    result = environment.probe(["/sbin/mount"], timeout=10)
    if result is None or result.returncode != 0:
        return {}
    target = os.path.realpath(path)
    found = {}
    longest = -1
    for line in result.stdout.splitlines():
        source, separator, rest = line.partition(" on ")
        mount, _, options = rest.partition(" (")
        mount = mount.strip()
        if not separator or not mount or not under(target, mount):
            continue
        if len(mount) > longest:
            longest = len(mount)
            found = {
                "fstype": options.split(",")[0].strip(" )"),
                "source": source.strip(),
                "target": mount,
            }
    return found


def filesystem_of(path):
    if sys.platform == "darwin":
        return darwin_filesystem(path)
    if not shutil.which("findmnt"):
        return {}
    result = environment.probe(["findmnt", "-no", "FSTYPE,SOURCE,TARGET", "--target", path], 10)
    if result is None or result.returncode != 0:
        return {}
    fields = result.stdout.split()
    if len(fields) < 3:
        return {}
    return {"fstype": fields[0], "source": fields[1], "target": fields[2]}


def suite_name(module):
    return module.__name__.rsplit(".", 1)[-1]


def collect_jobs(setting):
    found = []
    failures = []
    for module in SUITES:
        try:
            found.extend(module.jobs(setting))
        except (MeasurementError, OSError, subprocess.SubprocessError, ValueError) as error:
            failures.append(f"{suite_name(module)}: {error}")
    if setting.families:
        found = [job for job in found if job.outputs and family_of(job) in setting.families]
    return found, tuple(failures)


def family_of(job):
    return job.outputs[0].key.split(".", 1)[0]


def converged(collected):
    if not collected:
        return False
    return all(relative_deviation(values) <= RSD_THRESHOLD for values in collected.values())


def measure_job(job):
    collected = {}
    minimum = MIN_RUNS if job.repeat else 1
    limit = MAX_RUNS if job.repeat else 1
    attempts = 0
    while attempts < limit:
        values = job.measure()
        attempts += 1
        for key, value in values.items():
            bucket = collected.setdefault(key, [])
            if isinstance(value, (list, tuple)):
                bucket.extend(float(item) for item in value)
            else:
                bucket.append(float(value))
        if attempts >= minimum and converged(collected):
            break
    return collected


def metrics_for(job, collected):
    found = []
    for output in job.outputs:
        samples = collected.get(output.key)
        if not samples:
            continue
        found.append(
            Metric(
                key=output.key,
                method=job.method,
                scale=output.scale,
                proportion=output.proportion,
                comparable=output.comparable,
                tool=job.tool,
                tool_version=job.version,
                samples=tuple(samples),
                detail=dict(job.detail),
            )
        )
    return found


def timestamp():
    return datetime.now(UTC).replace(microsecond=0)


def run_id_for(started, snapshot):
    return f"{started.strftime('%Y-%m-%dT%H-%M-%SZ')}-{epoch_of(snapshot)}"


def cool_down(workdir, report):
    deadline = time.monotonic() + COOLDOWN_SECONDS
    while time.monotonic() < deadline:
        if report:
            report("cool", "cooling", f"{int(deadline - time.monotonic())}s left")
        time.sleep(COOLDOWN_INTERVAL)
        snapshot = collect_snapshot(full=True)
        values = capture_conditions(snapshot, workdir)
        if not values["throttled_at_start"]:
            return snapshot, values
    return None, None


def settle(snapshot, conditions, workdir, force, report):
    if force or not conditions.get("throttled_at_start"):
        return snapshot, conditions
    cooled_snapshot, cooled = cool_down(workdir, report)
    if cooled is None:
        return snapshot, conditions
    return cooled_snapshot, cooled


def execute(host, tier, families=(), note="", tags=(), force=False, workdir="", report=None):
    workdir = workdir or default_workdir()
    setting = Setting(tier=tier, workdir=workdir, families=tuple(families))
    snapshot = collect_snapshot(full=True)
    conditions = capture_conditions(snapshot, workdir)
    snapshot, conditions = settle(snapshot, conditions, workdir, force, report)
    conditions["filesystem"] = filesystem_of(workdir)
    writes = WRITE_BUDGET.get(tier, 0) > 0
    reasons = gate_reasons(conditions, writes_disk=writes)
    if reasons and not force:
        raise GateError(reasons)
    described = environment.describe_snapshot(snapshot)
    started = timestamp()
    jobs, failures = collect_jobs(setting)
    failures = list(failures)
    budget = WRITE_BUDGET.get(tier, 0)
    metrics = []
    written = 0
    interrupted = False
    try:
        for position, job in enumerate(jobs, start=1):
            if job.writes and written + job.writes > budget:
                over = (written + job.writes) / 1024**3
                refused = f"would write {over:.1f} GiB, past the {tier} budget"
                failures.append(f"{job.name}: {refused}")
                if report:
                    report("skip", job.name, refused)
                continue
            if report:
                report("start", job.name, f"{position} of {len(jobs)}")
            began = time.monotonic()
            try:
                collected = measure_job(job)
            except (MeasurementError, OSError, subprocess.SubprocessError, ValueError) as error:
                failures.append(f"{job.name}: {error}")
                if report:
                    report("skip", job.name, str(error))
                continue
            # fio's own io_bytes excludes the ramp, so it under-reports; the
            # predicted figure is exact now that the write stages are size bounded.
            measured = int(max(collected.pop(WRITTEN, [0.0]), default=0.0))
            written += max(job.writes, measured)
            produced = metrics_for(job, collected)
            metrics.extend(produced)
            if report:
                samples = max((metric.times_to_run for metric in produced), default=0)
                report("done", job.name, f"{time.monotonic() - began:.0f}s, n={samples}")
    except KeyboardInterrupt:
        interrupted = True
    grade = "aborted" if interrupted else grade_for(reasons, metrics, failures)
    return Run(
        run_id=run_id_for(started, described),
        host=host,
        started=started.isoformat().replace("+00:00", "Z"),
        tier=tier,
        grade=grade,
        snapshot=described,
        install=environment.describe_install(snapshot),
        conditions=conditions,
        metrics=tuple(metrics),
        note=note,
        tags=tuple(tags),
        dotfiles_sha=environment.dotfiles_sha(),
        gate_reasons=tuple(reasons) + tuple(failures),
        bytes_written=written,
    )

File: scripts/rust/crates/hwtune/src/bench/runner.rs
use super::{
    capture, conditions,
    record::{Metric, Run, epoch_of, relative_deviation},
    suites::{self, Job, Samples, WRITTEN},
};
use chrono::{SecondsFormat, Utc};
use std::{
    fs,
    path::PathBuf,
    thread,
    time::{Duration, Instant},
};
pub fn cancelled() -> bool {
    ui_terminal::termination_requested() || crate::tune::monitor::cancelled()
}

pub const FAMILIES: [&str; 7] = ["cpu", "mem", "cache", "disk", "gpu", "thermal", "workload"];
#[derive(Clone, Debug)]
pub struct Setting {
    pub tier: String,
    pub workdir: PathBuf,
    pub families: Vec<String>,
    pub memory_bytes: u64,
}
impl Setting {
    pub fn accepts(&self, family: &str) -> bool {
        self.families.is_empty() || self.families.iter().any(|selected| selected == family)
    }
}
#[derive(Default)]
pub struct Options {
    pub context: Option<super::provenance::RunContext>,
    pub host: String,
    pub tier: String,
    pub families: Vec<String>,
    pub note: String,
    pub tags: Vec<String>,
    pub force: bool,
    pub workdir: Option<PathBuf>,
}
pub fn write_budget(tier: &str) -> u64 {
    match tier {
        "standard" => 30 * 1024_u64.pow(3),
        "heavy" => 70 * 1024_u64.pow(3),
        _ => 0,
    }
}
pub fn default_workdir() -> PathBuf {
    super::suites::gpu::cache_dir().join("work")
}
pub fn converged(collected: &Samples) -> bool {
    !collected.is_empty()
        && collected
            .values()
            .all(|values| relative_deviation(values) <= 2.5)
}
pub fn measure_job(job: &mut Job) -> Result<Samples, String> {
    let mut collected = Samples::new();
    let minimum = if job.repeat { 3 } else { 1 };
    let limit = if job.repeat { 6 } else { 1 };
    for attempt in 1..=limit {
        if cancelled() {
            return Err("benchmark interrupted".into());
        }
        let measured = (job.measure)()?;
        for (key, values) in measured.values {
            if values.iter().any(|value| !value.is_finite()) {
                return Err(format!("{} reported non-finite measurements", job.name));
            }
            collected.entry(key).or_default().extend(values);
        }
        if let Some(detail) = measured.detail.as_object() {
            for (key, value) in detail {
                job.detail[key] = value.clone();
            }
        }
        if attempt >= minimum && converged(&collected) {
            break;
        }
    }
    Ok(collected)
}
pub fn metrics_for(job: &Job, collected: &Samples) -> Vec<Metric> {
    job.outputs
        .iter()
        .filter_map(|output| {
            let samples = collected
                .get(&output.key)
                .filter(|values| !values.is_empty())?;
            let mut metric = output.clone();
            metric.method = job.method.clone();
            metric.tool = job.tool.clone();
            metric.tool_version = job.version.clone();
            metric.samples = samples.clone();
            metric.detail = job.detail.clone();
            Some(metric)
        })
        .collect()
}
pub fn execute(options: &Options, report: &mut dyn FnMut(&str, &str, &str)) -> Result<Run, String> {
    let workdir = options.workdir.clone().unwrap_or_else(default_workdir);
    fs::create_dir_all(&workdir).map_err(|error| format!("{}: {error}", workdir.display()))?;
    let mut snapshot = sysinfo::collect::collect_snapshot_for_host(
        true,
        &options.host,
        &super::hosts::inventory_context(),
    );
    let mut conditions = conditions::capture_conditions(&snapshot, &workdir);
    if !options.force && conditions["throttled_at_start"] == true {
        let deadline = Instant::now() + Duration::from_secs(240);
        while Instant::now() < deadline {
            report(
                "cool",
                "cooling",
                &format!(
                    "{}s left",
                    deadline.saturating_duration_since(Instant::now()).as_secs()
                ),
            );
            for _ in 0..150 {
                if cancelled() {
                    return Err("benchmark interrupted".into());
                }
                thread::sleep(Duration::from_millis(100));
            }
            let cooled = sysinfo::collect::collect_snapshot_for_host(
                true,
                &options.host,
                &super::hosts::inventory_context(),
            );
            let values = conditions::capture_conditions(&cooled, &workdir);
            if values["throttled_at_start"] != true {
                snapshot = cooled;
                conditions = values;
                break;
            }
        }
    }
    conditions["filesystem"] = capture::filesystem_of(&workdir);
    let budget = write_budget(&options.tier);
    let writes_disk = budget > 0
        && (options.families.is_empty() || options.families.iter().any(|family| family == "disk"));
    let reasons = conditions::gate_reasons(&conditions, writes_disk);
    if !reasons.is_empty() && !options.force {
        return Err(format!(
            "{}\nconditions are not suitable; pass --force to measure anyway",
            reasons.join("\n")
        ));
    }
    let described = capture::describe_snapshot(&snapshot);
    let started = Utc::now();
    let setting = Setting {
        tier: options.tier.clone(),
        workdir,
        families: options.families.clone(),
        memory_bytes: snapshot.result("Memory")["total"].as_u64().unwrap_or(0),
    };
    let (mut jobs, mut failures) = suites::collect_jobs(&setting);
    let count = jobs.len();
    let mut metrics = Vec::new();
    let mut written = 0_u64;
    for (position, job) in jobs.iter_mut().enumerate() {
        if cancelled() {
            break;
        }
        if job.writes > 0 && written.saturating_add(job.writes) > budget {
            let refused = format!(
                "would write {:.1} GiB, past the {} budget",
                written.saturating_add(job.writes) as f64 / 1073741824.0,
                options.tier
            );
            failures.push(format!("{}: {refused}", job.name));
            report("skip", &job.name, &refused);
            continue;
        }
        report("start", &job.name, &format!("{} of {count}", position + 1));
        // Reserve before invoking a tool: failed/aborted jobs may have written
        // their files already and must still consume this run's budget.
        written = written.saturating_add(job.writes);
        let began = Instant::now();
        let mut collected = match measure_job(job) {
            Ok(values) => values,
            Err(error) => {
                failures.push(format!("{}: {error}", job.name));
                report("skip", &job.name, &error);
                continue;
            }
        };
        let measured = collected
            .remove(WRITTEN)
            .unwrap_or_default()
            .into_iter()
            .fold(0.0, f64::max) as u64;
        written = written.saturating_add(measured.saturating_sub(job.writes));
        let produced = metrics_for(job, &collected);
        let samples = produced.iter().map(|m| m.samples.len()).max().unwrap_or(0);
        metrics.extend(produced);
        report(
            "done",
            &job.name,
            &format!("{}s, n={samples}", began.elapsed().as_secs()),
        );
    }
    let grade = if cancelled() {
        "aborted"
    } else {
        conditions::grade_for(&reasons, metrics.len(), &failures)
    };
    let gate_reasons = reasons.into_iter().chain(failures).collect();
    Ok(Run {
        run_id: format!(
            "{}-{}-{:x}",
            started.format("%Y-%m-%dT%H-%M-%S%.9fZ"),
            epoch_of(&described),
            std::process::id()
        ),
        context: options.context.clone(),
        host: options.host.clone(),
        started: started.to_rfc3339_opts(SecondsFormat::Secs, true),
        tier: options.tier.clone(),
        grade: grade.into(),
        snapshot: described,
        install: sysinfo::report::describe_install(&snapshot),
        conditions,
        metrics,
        note: options.note.clone(),
        tags: options.tags.clone(),
        dotfiles_sha: capture::dotfiles_sha(),
        gate_reasons,
        bytes_written: written,
        schema: 1,
    })
}
